src/auth/auth_gitlab.py

```diff
@@ -0,0 +1,339 @@
+class GitlabOAuthAuthenticator(auth_base.Authenticator, GitlabOAuth2Mixin):
+    def __init__(self, params_dict):
+        super().__init__()
+
+        LOGGER.debug("Init gitlab oauth provider with " + str(params_dict))
+
+        self.client_id = model_helper.read_obligatory(params_dict, 'client_id', ' for Gitlab OAuth')
+
+        secret_value = model_helper.read_obligatory(params_dict, 'secret', ' for Gitlab OAuth')
+        self.secret = model_helper.resolve_env_vars(secret_value, full_match=True)
+
+        gitlabPrefix = params_dict.get('url')
+        if not model_helper.is_empty(gitlabPrefix):
+            self._GITLAB_PREFIX = gitlabPrefix
+
+        self.states = {}
+        self.user_states = {}
+        self.gitlab_update = params_dict.get('auth_info_ttl')
+        self.gitlab_dump = params_dict.get('state_dump_file')
+        self.gitlab_group_support = params_dict.get('group_support', True)
+        self.session_expire = int(params_dict.get('session_expire_minutes', 0)) * 60
+        now = time.time()
+
+        if self.gitlab_dump and os.path.exists(self.gitlab_dump):
+            dumpFile = open(self.gitlab_dump, "r")
+            stateStr = dumpFile.read()
+            self.user_states = escape.json_decode(stateStr)
+            dumpFile.close()
+            for userData in list(self.user_states.keys()):
+                # force to update user from gitlab
+                self.user_states[userData]['updating'] = False
+                if self.gitlab_update:
+                    self.user_states[userData]['updated'] = now - self.gitlab_update - 1
+            LOGGER.info("Readed state from file %s: " % self.gitlab_dump + str(self.user_states))
+
+        self.gitlab_group_search = params_dict.get('group_search')
+
+        self._client_visible_config['client_id'] = self.client_id
+        self._client_visible_config['oauth_url'] = self._OAUTH_AUTHORIZE_URL % self._GITLAB_PREFIX
+        self._client_visible_config['oauth_scope'] = 'api' if self.gitlab_group_support else 'read_user'
+
+    def authenticate(self, request_handler):
+        code = request_handler.get_argument('code', False)
+
+        if not code:
+            LOGGER.error('Code is not specified')
+            raise AuthBadRequestException('Missing authorization information. Please contact your administrator')
+
+        return self.validate_user(code, request_handler)
+
+    def is_active(self, user, request_handler):
+        access_token = request_handler.get_secure_cookie('token')
+        if access_token is None:
+            return False
+        access_token = access_token.decode("utf-8")
+
+        if self.user_states.get(user) is None:
+            LOGGER.debug("User %s not found in state" % user)
+            return False
+
+        if self.user_states[user]['state'] is None or self.user_states[user]['state'] != "active":
+            LOGGER.info("User %s state inactive: " % user + str(self.user_states[user]))
+            del self.user_states[user]
```

Comment:
I would suggest, that reading method (as for me, `is_smth` is a reading method) shouldn't have any side effects and modify any data. What do you think?
Also, if the user is not active, we are doing logout anyway

Maybe we could just use another name. Actually I think "validate" user might be better fit here :)

Reply:
Yep, I will rename it to validate or validate_user